js-envs/src/main/scala/org/scalajs/jsenv/nodejs/JSDOMNodeJSEnv.scala

```diff
@@ -0,0 +1,110 @@
+      code: VirtualJSFile): JSRunner = {
+    new DOMNodeRunner(libs, code)
+  }
+
+  override def asyncRunner(libs: Seq[ResolvedJSDependency],
+      code: VirtualJSFile): AsyncJSRunner = {
+    new AsyncDOMNodeRunner(libs, code)
+  }
+
+  override def comRunner(libs: Seq[ResolvedJSDependency],
+      code: VirtualJSFile): ComJSRunner = {
+    new ComDOMNodeRunner(libs, code)
+  }
+
+  protected class DOMNodeRunner(libs: Seq[ResolvedJSDependency], code: VirtualJSFile)
+      extends ExtRunner(libs, code) with AbstractDOMNodeRunner
+
+  protected class AsyncDOMNodeRunner(libs: Seq[ResolvedJSDependency], code: VirtualJSFile)
+      extends AsyncExtRunner(libs, code) with AbstractDOMNodeRunner
+
+  protected class ComDOMNodeRunner(libs: Seq[ResolvedJSDependency], code: VirtualJSFile)
+      extends AsyncDOMNodeRunner(libs, code) with NodeComJSRunner {
+    override protected def initFiles(): Seq[VirtualJSFile] =
+      super.initFiles ++ comSetup
+  }
+
+  protected trait AbstractDOMNodeRunner extends AbstractNodeRunner {
+
+    protected def codeWithJSDOMContext(): Seq[VirtualJSFile] = {
+      val scriptsJSPaths = getLibJSFiles().map {
+        case file: FileVirtualFile => file.path
+        case file                  => libCache.materialize(file).getAbsolutePath
+      }
+      val scriptsStringPath = scriptsJSPaths.map('"' + _ + '"')
+      val jsDOMCode = {
+        s"""
+           |(function () {
+           |  const jsdom = require("jsdom");
+           |  var windowKeys = [];
+           |
+           |  jsdom.env({
+           |    html: "",
+           |    virtualConsole: jsdom.createVirtualConsole().sendTo(console),
+           |    created: function (error, window) {
+           |      if (error == null) {
+           |        window["__ScalaJSEnv"] = __ScalaJSEnv;
+           |        window["scalajsCom"] = global.scalajsCom;
+           |        windowKeys = Object.keys(window);
+           |      } else {
+           |        console.log(error);
+           |      }
+           |    },
+           |    scripts: [${scriptsStringPath.mkString(", ")}],
+           |    onload: function (window) {
+           |      for (k in window) {
+           |        if (windowKeys.indexOf(k) == -1)
+           |          global[k] = window[k];
+           |      }
+           |
+           |      ${code.content}
+           |    }
+           |  });
+           |})();
+           |""".stripMargin
+      }
+      Seq(new MemVirtualJSFile("codeWithJSDOMContext.js").withContent(jsDOMCode))
+    }
+
+    override protected def getJSFiles(): Seq[VirtualJSFile] =
+      initFiles() ++ codeWithJSDOMContext()
+
+    /** Libraries are loaded via scripts in Node.js */
+    override protected def getLibJSFiles(): Seq[VirtualJSFile] = {
+      customInitFiles() ++
+      libs.map(_.lib)
+    }
+
+    // Send code to Stdin
+    override protected def sendVMStdin(out: OutputStream): Unit = {
+      sendJS(getJSFiles(), out)
+    }
```

Comment:
Both `AbstractDOMNodeRunner` and `AbstractBasicNodeRunner` define this method in the same way. Why is not defined in `AbstractNodeRunner` instead?